src/content/blog/astro-4100.mdx

```diff
@@ -0,0 +1,146 @@
+```
+
+## Experimental: astro:env
+
+Astro 4.10 introduces a new experimental built-in module, `astro:env`.
+
+Environment variables allow you to configure your app with different values in different environments. But this comes with a great deal of complexity:
+
+- Some variables are used in the client and some only on the server.
+- Server variables are often __secrets__, things like API keys that you absolutely do not want exposed in the client, and you also do not want inlined into the server build either (which can be viewed by anyone with access to the build output).
+- Some variables are __required__ for your app to function at all; whereas other are optional enhancements.
+- Runtimes such as Cloudflare and Deno have different APIs for reading variables, creating a dev/prod difference you need to deal with.
+
+We built `astro:env` to provide more control and structure over environment variables. Here's how you might define a schema for them in your config:
+
+```js
+import { defineConfig, envField } from 'astro/config';
+
+export default defineConfig({
+  env: {
+    schema: {
+      DATABASE_URL: envField.string({
+        context: 'client',
+        access: 'public'
+      }),
+      API_PORT: envField.number({
+        context: 'server',
+        access: 'secret',
+        default: 7000
+      })
+    }
+  }
+})
+```
+
+Once defined you can use your variables by importing them from the `astro:env` module like so:
+
+```js
+import { API_PORT } from 'astro:env/server';
+
+// API_PORT is a number
+// ...
+```
+
+Some times you still want to read a variable that is not defined in the schema. This might be because the variable name is dynamic itself such as `BACK_SERVER_1` and `BACK_SERVER_2`, for example. Or it might be because you are reading variables inside of an Astro integration and don't want to impose the schema onto the app. In these cases `getSecret` is available to read runtime variables and works with any runtime (Cloudflare, Node.js, Deno).
+
+```js
+import { getSecret } from 'astro:env/server';
+
+function getServerEndpoint(num: number) {
+  return getSecret('BACKUP_SERVER_' + num);
+}
+```
+
+## Rewrite for all HTTP methods
+
+Rewrites is a new experimental feature released in 4.9. The first version targeted `GET` requests, the most common case for a rewrite. Now in 4.10 rewrites can be used to change the route for any requests, which it does by cloning the initial request.
+
+A use-case would be having a versioned API with a default version that you rewrite in middleware:
+
+```js
+import { defineMiddleware } from 'astro:middleware';
+
+// `context` and `next` are automatically typed
+export const onRequest = defineMiddleware(({ request, url }, next) => {
+  if(request.method === 'POST' && url.pathname === '/api') {
+    return next('/api/v2');
+  }
+});
+```
+
+When rewriting a new request is created pointing to the new URL. The existing headers and body are copied over to the new request.
+
+## Embedding Astro
```

Comment:
Maybe we could add a new chapter about people should load renderers, this changed now: https://github.com/withastro/astro/blob/feat/load-renderers/.changeset/fair-singers-reflect.md

Reply:
Sarah beat me to it.